make WC_Data_Store update_lookup_table() public

Jon007 opened the pull request:
make update_lookup_table public so it can be called to force updates, for example for compatibility with legacy plugins / other which make updates to post meta leaving the lookup table out of date.

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->
Changes WC_Data_Store_WP update_lookup_table() to public
The reason is to allow resynchronisation of lookup table in cases where it may become out of date, for example where underlying meta data may have been changed by legacy plugins using post meta api rather than product api.

### How to test the changes in this Pull Request:

1. code example
$data_store = \WC_Data_Store::load( $datastoreType );
$data_store->update_lookup_table( $productid, 'wc_product_meta_lookup' );

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.
make update_lookup_table public() to allow resynchronisation of lookup table data by other plugins

Jon007 committed:
make update_lookup_table public to force update

make update_lookup_table public so it can be called to force updates, for example for compatibility with legacy plugins / other which make updates to post meta leaving the lookup table out of date.

Jon007 commented:
here is an example usage:

```
	/*
	 * Ensure product cache is refreshed and product_meta_lookup updated for any changed meta
	 *
	 * @param int $id   id of Product
	 */
	public static function flushCacheUpdateLookupTable( $id ) {
		$product = wc_get_product( $id );
		if ( $product ) {
			wc_delete_product_transients( $id );
			wp_cache_delete( $id, 'post_meta' );
			wp_cache_delete( 'lookup_table', 'object_' . $id );
			$productType = $product->get_type();

			$datastoreType = 'product';
			switch ( $productType ) {
				case 'variable':
				case 'grouped':
				case 'variation':
					$datastoreType .= '-' . $productType;
			}
			$data_store = \WC_Data_Store::load( $datastoreType );
			//not currently allowed as protected function
			//$data_store->update_lookup_table( $id, 'wc_product_meta_lookup' );
			//in the meantime an increase of zero in the product sales will force the update...
			$data_store->update_product_sales( $id, 0, 'increase' );
		}
	}
```
I welcome any alternatives and apologise in advance if I have missed anything that already exists - if there isn't already a function for this it would also be useful.  Yes I am aware that ideally all plugins should only use product api, but that is a way off yet and this offers a solution in the meantime.

Jon007 committed:
rating optional = generate_product_data incorrect

If rating is not mandatory on comments, generate_product_data() produces incorrect output.
For example items with comments but no reviews are output with rating of zero which is getting flagged in google webmasters tools under the heading "Rating is missing required best and/or worst values".

The output google is objecting to is eg:
"aggregateRating":{"@type":"AggregateRating","ratingValue":"0","reviewCount":1}}

postmeta in this case is:
_wc_average_rating 0
_wc_rating_count 0.0
_wc_review_count 1

product lookup table is reporting
rating_count 0
average_rating 0.00

the meta is correct and reflected in the product object, it is the test that is wrong since
if $product->get_review_count() 
may be true even if there are no ratings.

claudiosanches commented on includes/class-wc-structured-data.php:
```diff
@@ -268,7 +268,7 @@ public function generate_product_data( $product = null ) {
 		}
 
-		if ( $product->get_review_count() && wc_review_ratings_enabled() ) {
+		if ( $product->get_rating_count() && wc_review_ratings_enabled() ) {
```
Seems like that you included some extra code in this PR, if you could please revert it.

Jon007 replied:
Hi, yes, this is an unrelated fix:

If rating is not mandatory on comments, generate_product_data() produces incorrect output.
For example items with comments but no reviews are output with rating of zero which is getting flagged in google webmasters tools under the heading "Rating is missing required best and/or worst values".

The output google is objecting to is eg:
"aggregateRating":{"@type":"AggregateRating","ratingValue":"0","reviewCount":1}}

postmeta in this case is:
_wc_average_rating 0
_wc_rating_count 0.0
_wc_review_count 1

product lookup table is reporting
rating_count 0
average_rating 0.00

the meta is correct and reflected in the product object, it is the test that is wrong since
if $product->get_review_count() 
may be true even if there are no ratings.
therefore the correct code should be to use get_rating_count() instead

Jon007 replied:
ok, removed from this pull

claudiosanches reviewed (changes requested):
This change looks fine, but seems like you included some extra code in here.

Jon007 committed:
reverse  to move to separate pull request

get_review_count() / get_rating_count() issue on generate_product_data() to be submitted separately

claudiosanches reviewed (approved):
LGTM!

cc @mikejolley

claudiosanches commented:
Putting back to review, since we need to validate if this should not cause any trouble.

Jon007 commented:
Could we add a wc_refresh_product($the_product) to wc-product-functions.php ?

For example a 3rd party would currently have to implement something like this to test if update_lookup_table is public yet, whereas testing for existence of a wc_refresh_product() or similar and calling it would be much simpler. 

```
	/*
	 * Ensure product cache is refreshed and product_meta_lookup updated for any changed meta
	 *
	 * @param int $id   id of Product
	 */
	public static function wc_refresh_product( $the_product ) {
		$product = wc_get_product( $the_product );
		if ( $product ) {
			$id			 = $product->get_id();
			wc_delete_product_transients( $id );
			wp_cache_delete( $id, 'post_meta' );
			wp_cache_delete( 'lookup_table', 'object_' . $id );
			$productType = $product->get_type();

			$datastoreType = 'product';
			switch ( $productType ) {
				case 'variable':
				case 'grouped':
				case 'variation':
					$datastoreType .= '-' . $productType;
			}
			$data_store = \WC_Data_Store::load( $datastoreType );
			//woocommerce 3.6 initial releases made all the lookup table functions protected so we cannot call them
			//note that is_callable() cannot test for implementation here as root class WC_Data_Store implements __call()
			//if ( is_callable( array( $data_store, 'update_lookup_table' ) ) ) {
			//method_exists also returns true if protected so also need reflection to confirm if it is callable
			//cant test this on the datastore object itself as the method is not part of the datastore interface
			//so test the product data store itself
			if ( method_exists( 'WC_Product_Data_Store_CPT', 'update_lookup_table' ) ) {
				$product_data_store	 = new \WC_Product_Data_Store_CPT();
				$reflection			 = new \ReflectionMethod( $product_data_store, 'update_lookup_table' );
				if ( $reflection->isPublic() ) {
					$data_store->update_lookup_table( $id, 'wc_product_meta_lookup' );
				} else {
					//in the meantime an increase of zero in the product sales will force the update...
					$data_store->update_product_sales( $id, 0, 'increase' );
				}
			}
		}
	}

```

kloon commented:
I had a chat with Mike around this PR and we both feel this open up a whole for plugins/3rd party code to work around CRUD. It is also easy for the data to go out of sync with what is on the product if we make this public.

CRUD has been in core for quite a while now so there is no excuse for plugins and third-party code to not not make use of it, I would argue it would perhaps be less effort in converting to CRUD than it would be to manually make calls to this function if public.

With that said, I am going to close this PR without merging. Sorry, but hope you understand the reasoning.

Jon007 commented:
@kloon I do understand that woocommerce specific plugins should be aware of CRUD by now, 

However please consider that there are a lot of non-woocommerce plugins dealing with common features of posts such as featured images and cloning/copying/synchronising posts with their postmeta etc.

The effect of not being able to force synchronise the lookup table is that there is no ready workaround for compatibility for such generic plugins.

In fact such plugins will continue to work, but with only subtle and not-readily-fixable flaws which only the cautious tester will notice.

A standard woocommerce api to notify woocommerce that something wordpress-y has happened would remediate that.
